appengine/pond_storage/rank.py

```diff
@@ -15,72 +15,165 @@
 limitations under the License.
 """
 
-"""Return a ranking match request.
-"""
-
-__author__ = "kozbial@google.com (Monica Kozbial)"
+"""Returns a ranking match request or processes the given match results."""
 
+import cgi
 import datetime
 import json
+import logging
 from google.appengine.ext import ndb
 from pond_storage import *
 
-"""Store match request information."""
 class MatchRequest(ndb.Model):
-  # Ducks.
-  ducks = ndb.KeyProperty(repeated=True)
-  # Creation time.
+  """Stores match request information."""
+  entry_keys = ndb.KeyProperty(repeated=True, kind=LeaderboardEntry)
   creation_time = ndb.DateTimeProperty(auto_now_add=True)
-  # Expiry time.
-  expiry_time = ndb.ComputedProperty(lambda self: self.creation_time + datetime.timedelta(hours=1))
+  expiry_time = ndb.ComputedProperty(
+      lambda self: self.creation_time + datetime.timedelta(minutes=1))
 
   @classmethod
-  def contains_duck(cls, duck_key):
-    return cls.query(cls.ducks == duck_key).count(limit=1)
-
-# Cleanup expired match requests.
-now = datetime.datetime.now()
-expired_requests = MatchRequest.query(MatchRequest.expiry_time < now)
-for request in expired_requests:
-  request.key.delete()
-
-# Create a new match request.
-# 1. Choose duck with highest instability not already in a match request.
-target_duck = None
-entries_query = LeaderboardEntry.query().order(-LeaderboardEntry.instability)
-for entry in entries_query:
-  # Check if it exists in a current match request
-  if not MatchRequest.contains_duck(entry.duck_key):
-    target_duck = entry.duck_key.get()
-    if target_duck:
+  def contains_entry(cls, entry_key):
+    return cls.query(cls.entry_keys == entry_key).count(limit=1)
+
+def create_ranking_match():
+  """Creates a new ranking match if a valid match can be created.
+
+  Chooses the leaderboard entry with the highest instability no already being
+  re-evaluated in a ranking match and finds three more entries relative to it
+  are also not involved in a pending match. If four valid entries can be found,
```

Comment:
nit: Chooses the leaderboard entry with the highest instability that is not in a pending match and finds three more entries relative to it that are also not in a pending match.

Reply:
Updated.